src/Http/Http.Extensions/src/HttpContextJsonExtensions.Response.cs

```diff
@@ -0,0 +1,80 @@
+
+using System;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http.Json
+{
+    public static partial class HttpContextJsonExtensions
+    {
+        public static ValueTask WriteAsJsonAsync<TValue>(
+            this HttpResponse response,
+            TValue value,
+            CancellationToken cancellationToken = default)
+        {
+            return response.WriteAsJsonAsync<TValue>(value, options: null, contentType: JsonConstants.JsonContentType, cancellationToken);
+        }
+
+        public static ValueTask WriteAsJsonAsync<TValue>(
+            this HttpResponse response,
+            TValue value,
+            JsonSerializerOptions options,
+            CancellationToken cancellationToken = default)
+        {
+            return response.WriteAsJsonAsync<TValue>(value, options, contentType: JsonConstants.JsonContentType, cancellationToken);
+        }
+
+        public static ValueTask WriteAsJsonAsync<TValue>(
+            this HttpResponse response,
+            TValue value,
+            JsonSerializerOptions options,
+            string contentType,
+            CancellationToken cancellationToken = default)
+        {
+            if (contentType != null)
+            {
+                response.ContentType = contentType;
+            }
+            response.StatusCode = StatusCodes.Status200OK;
+            return new ValueTask(JsonSerializer.SerializeAsync<TValue>(response.Body, value, options, cancellationToken));
+        }
+
+        public static ValueTask WriteAsJsonAsync(
+            this HttpResponse response,
+            Type type,
+            object value,
+            CancellationToken cancellationToken = default)
+        {
+            return response.WriteAsJsonAsync(type, value, options: null, contentType: JsonConstants.JsonContentType, cancellationToken);
+        }
+
+        public static ValueTask WriteAsJsonAsync(
+            this HttpResponse response,
+            Type type,
+            object value,
+            JsonSerializerOptions options,
+            CancellationToken cancellationToken = default)
+        {
+            return response.WriteAsJsonAsync(type, value, options, contentType: JsonConstants.JsonContentType, cancellationToken);
+        }
+
+        public static ValueTask WriteAsJsonAsync(
+            this HttpResponse response,
+            Type type,
+            object value,
+            JsonSerializerOptions options,
+            string contentType,
+            CancellationToken cancellationToken = default)
+        {
+            if (contentType != null)
+            {
+                response.ContentType = contentType;
+            }
+            response.StatusCode = StatusCodes.Status200OK;
+            return new ValueTask(JsonSerializer.SerializeAsync(response.Body, value, type, options, cancellationToken));
```

Comment:
Same thing as reading. Would need to use a transcoder if the charset is not utf-8

Reply:
We intentionally only support utf-8.

Reply:
> Replace the guts of MVC's JSON formatter with this (validates flexibility)

That's not going to happen if this does not support utf8. Was this made a non-goal?

Reply:
I'm with Pranav, we have a new type in the BCL to make this trivial that we also use in the HttpClient versions of this API. We should do the same and not restrict this to UTF8 only.

Reply:
I agree. I believe we limited to UTF8 for simplicity and to reduce overloads on `WriteAsJsonAsync`.

`ReadFromJsonAsync`: Simple to update it to look at request data and make decisions
`WriteAsJsonAsync`: Pranav mentioned that we could get the charset from the content-type.

`WriteAsJsonAsync(o, null, contentType: null)` = UTF8
`WriteAsJsonAsync(o, null, contentType: "application/json")` = UTF8
`WriteAsJsonAsync(o, null, contentType: "application/json; charset=utf-8")` = UTF8
`WriteAsJsonAsync(o, null, contentType: "application/json; charset=utf-16")` = UTF16

Thoughts?

Reply:
@pranavkm What does MVC do if you specify a charset that it doesn't know about? I think MVC has configuration to allow you to specify new encodings. This helper method wouldn't have that.

Reply:
A major point from the design review was not trying to parse the response content-type to determine Write behavior, the content-type is always pass-through. If you want to support an alternate encoding then that should be passed in as a its own strongly typed parameter.

The other point was that the need to write (as opposed to read) alternate encodings is rare enough that it didn't warrant support on these convenience APIs, you can use the serializer directly.

It's different for the request were the content-type charset is the only way to know the correct encoding to read.

Reply:
There is a conflict between the API surface as currently proposed, and requirement to use in MVC.

MVC has configuration to specify additional encodings. MVC would need to be able to resolve the content-type to an encoding and then pass that to ReadFromJsonAsync and WriteFromJsonAsync in its formatters.

I see some different options:
1. Leave ReadFromJsonAsync and WriteFromJsonAsync as UTF8 only
2. Support UTF8 and UTF16 in ReadFromJsonAsync automatically (similar to MVC's default config). Leave WriteFromJsonAsync unchanged.
3. Add overloads with `Encoding` parameter to read and write methods. Methods would be usable by MVC formatters.

Reply:
So we'll always write utf8 but support any encoding when reading?

Reply:
Which option are you talking about?

Reply:
Honestly we should re-think the MVC requirement, it conflicts with the main goal to "Make this the easiest imperative way to read/write JSON". MVC is fine using the lower level APIs.

> 2\. Support UTF8 and UTF16 in ReadFromJsonAsync automatically (similar to MVC's default config). Leave WriteFromJsonAsync unchanged.

This seems like the place to start, flexible on input and strict on output. We can add more overloads later if needed.  If we add overloads to specify an encoding it should only be on Write, read should only pivot on charset.

Reply:
Yea that seems fine. I don't like pivoting on charset for write either.